File: Core-backend-layer/Programming-fundamentals/Java/OOP/src/AbstractClasses/products/Product.java
package products;

public abstract class Product {
    String name;
    double price;
    private double costPrice; // Private field - only accessible within the same class
    protected int productId; // Protected field - accessible within the same class, within subclasses, and within the same package

    public Product(int productId, String name, double costPrice) {
        this.productId = productId;
        this.name = name;
        this.costPrice = costPrice;
    }

    // Abstract method for displaying information
    public abstract void displayInfo();

    // Concrete method
    public void setPrice(double newPrice) {
        this.price = newPrice;
        System.out.println(name + "'s price has been updated to $" + newPrice);
    }

    // Final method - cannot be overridden by subclasses
    public final double calculateTax() {
        return price * 0.17;
    }

    // Private method - only accessible within the same class
    private double calculateTotalCost() {
        return price * 0.05 + costPrice;
    }

    // Protected method - accessible within the same class, its subclasses, and classes in the same package
    protected double calculateProfit() {
        return price - calculateTotalCost() - calculateTax();
    }
}

File: Core-backend-layer/Programming-fundamentals/Python/OOP/src/abstract_classes/products/product.py
from abc import ABC, abstractmethod

class Product(ABC):
    def __init__(self, product_id, name, cost_price):
        self._product_id = product_id # Single leading underscore indicates that the variable is intended for internal use within the module or class
        self.name = name
        self.__cost_price = cost_price # Double leading underscore indicates that the variable is meant for internal use within the class and its subclasses
        self.price = 0  # Initializing to 0 is similar to Java not initializing in the constructor

    # Abstract method for displaying information
    @abstractmethod
    def display_info(self):
        pass

    # Concrete method
    def set_price(self, new_price):
        self.price = new_price
        print(f"{self.name}'s price has been updated to ${new_price}")

    # Python does not have a modifier to declare a method as final.
    def calculate_tax(self):
        return self.price * 0.17

    # Double leading underscore indicates that the method is meant for internal use within the class and its subclasses
    def __calculate_total_cost(self):
        return self.price * 0.05 + self.__cost_price

    # Single leading underscore indicates that the method is intended for internal use within the module or class
    def _calculate_profit(self):
        return self.price - self.__calculate_total_cost() - self.calculate_tax()